runtimes/c/runtime.c

```diff
@@ -11,87 +15,938 @@ typedef enum catala_fatal_error_code
+  const void* payload;
+} catala_option;
+
+#define CATALA_OPTION(_) catala_option*
+
+const catala_option catala_none = {catala_option_none, NULL};
+
+#define CATALA_NONE &catala_none
+
+const CATALA_OPTION(X) catala_some (const void* x) {
+  catala_option* ret;
+  ret = catala_malloc(sizeof(catala_option));
+  ret->code = catala_option_some;
+  ret->payload = x;
+  return ret;
+}
+
+CATALA_BOOL catala_isnone (const CATALA_OPTION() opt)
+{
+  return CATALA_NEW_BOOL(opt->code == catala_option_none);
+}
+
+const CATALA_OPTION() handle_exceptions
+  (const catala_code_position* pos,
+   const CATALA_ARRAY(const CATALA_OPTION()) e)
+{
+  int i, j;
+  unsigned int size = e->size;
+  const CATALA_OPTION() * excs = (const CATALA_OPTION() *)e->elements;
+  for (i = 0; i < size && excs[i]->code == catala_option_none; i++) {}
+  if (i >= size) return CATALA_NONE;
+  for(j = i + 1; j < size && excs[j]->code == catala_option_none; j++) {}
+  if (j < size) catala_error(catala_conflict, pos);
+  return excs[i];
+}
+
+void catala_init()
+{
+  mp_set_memory_functions(&catala_malloc,&catala_realloc,&catala_free);
+  mpz_init_set_ui(zconst_100, 100);
+  if (setjmp(catala_error_jump_buffer)) {
+    char *error_kind;
+    const catala_code_position pos = catala_error_raised.position;
+    switch (catala_error_raised.code) {
+    case catala_assertion_failed:
+      error_kind = "Asssertion failure";
+      break;
+    case catala_no_value:
+      error_kind = "No value provided";
+      break;
+    case catala_conflict:
+      error_kind = "Conflict between exceptions";
+      break;
+    case catala_division_by_zero:
+      error_kind = "Division by zero";
+      break;
+    case catala_not_same_length:
+      error_kind = "List lengths not matching";
+      break;
+    case catala_uncomparable_durations:
+      error_kind = "Comparison between incompatible durations";
+      break;
+    case catala_indivisible_durations:
+      error_kind = "Division of incompatible durations";
+      break;
+    case catala_malloc_error:
+      error_kind = "Out of memory";
+      break;
     }
+    if (pos.filename == NULL)
+      printf("\033[1;31m[ERROR]\033[m %s\n", error_kind);
+    else
+      printf("\033[1;31m[ERROR]\033[m %s in file %s:%d.%d-%d.%d\n",
+             error_kind,
+             pos.filename,
+             pos.start_line,
+             pos.start_column,
+             pos.end_line,
+             pos.end_column);
+    catala_free_all();
+    exit(10);
+  }
+  return;
 }
+
+/* void test()
+ * {
+ *   CATALA_MONEY dollars = catala_new_money_str ("10000");
+ *   CATALA_DEC rate = catala_new_dec_str ("1/3");
+ *   CATALA_DEC rate2 = catala_new_dec_str ("30/100");
+ *   CATALA_DEC ratex = o_add_rat_rat(rate, rate2);
+ *   CATALA_MONEY result = o_mult_mon_rat(dollars,ratex);
+ *   mpz_out_str(NULL,10,result);
+ *   printf("\n");
+ *   fflush(stdout);
+ * }
+ * 
+ * int main()
+ * {
+ *   catala_init();
+ *   test();
+ *   catala_free_all();
+ *   return 0;
+ * } */
```

Comment:
?

Reply:
This was a very basic unit test for the runtime; it's probably worth keeping, but should be integrated in the build system.